python/test/unit/runtime/test_autotuner.py

```diff
@@ -43,6 +43,51 @@ def _kernel(src, N, BLOCK_SIZE: tl.constexpr):
     triton.testing.assert_close(src, torch.ones_like(src))
 
 
+def test_hooks():
+    # Autotuner's pre- and post- hooks should be called the same number of times
+    N = 4096
+    src = torch.zeros(N, device='cuda')
+
+    configs = [triton.Config(kwargs={'BLOCK_SIZE': 4096}), triton.Config(kwargs={'BLOCK_SIZE': 32})]
+
+    @triton.autotune(configs=configs, key=['N'], warmup=1, rep=1)
+    @triton.heuristics({"N_STAGES": lambda nargs: 100 if nargs['N'] == 4096 else 4})
+    @triton.jit
+    def _kernel(src, N, N_STAGES: tl.constexpr, BLOCK_SIZE: tl.constexpr):
+        offsets = tl.arange(0, BLOCK_SIZE)
+        max_iters = tl.cdiv(N, BLOCK_SIZE)
+        for _ in tl.range(max_iters, num_stages=N_STAGES):
+            x = tl.load(src + offsets, mask=offsets < N)
+            tl.store(src + offsets, x, mask=offsets < N)
+            offsets += BLOCK_SIZE
+
+    values = {"counter": 0, "has_exception": False}
```

Comment:
Can these be two regular variables, instead of a dict?  I think you just need `global counter` inside the lambdas.

Reply:
I think in that case `counter` has to be a real global counter outside of the function scope, but we probably prefer to keep it inside the pytest function.

Reply:
Oh I didn't realize that.  SG then.